cmd/ndm_daemonset/probe/eventhandler.go

```diff
@@ -47,13 +51,53 @@ func (pe *ProbeEvent) addBlockDeviceEvent(msg controller.EventMessage) {
 	}
 	// iterate through each block device and perform the add/update operation
 	for _, device := range msg.Devices {
-		klog.Infof("Processing details for %s", device.UUID)
+		klog.Infof("Processing details for %s", device.DevPath)
 		pe.Controller.FillBlockDeviceDetails(device)
 		// if ApplyFilter returns true then we process the event further
 		if !pe.Controller.ApplyFilter(device) {
 			continue
 		}
-		klog.Infof("Processed details for %s : %s", device.DevPath, device.UUID)
+		klog.Infof("Processed details for %s", device.DevPath)
+
+		// if GPTBasedUUID need to be used, generate the UUID,
+		// if UUID cannot be generated create a GPT partition
+		if pe.Controller.FeatureGates.IsEnabled(features.GPTBasedUUID) {
+			if len(device.Partitions) > 0 {
+				klog.Info("device has partitions. not creating blockdevice resource")
+				continue
+			}
+			if len(device.Holders) > 0 {
+				klog.Info("device has holder devices, not creating blockdevice resource")
+				continue
+			}
+			uuid, ok := generateUUID(*device)
+			// manaully create a single partition on the device
+			if !ok {
```

Comment:
There is no need to create a partition -- creating a GPT label should be fine. The purpose is to ID the disk if we don't have HW identifiers and there are no existing GPT labels or FS ID's, right?

Reply:
Yes, the purpose is to ID the disk which do not have hardware identifiers. But, if we just write the GPT label to a disk suppose `/dev/sda`, then the consumer of that device can still overwrite the GPT label that we have written and we loose the unique identifier.

If we write the GPT label and then create a single partition that spans the entire drive, the consumers will be made available only the partition `/dev/sda1`, and our  GPT label will remain safe from overwrites.

Reply:
That can always happen and you have no way of avoiding that. If you "fdisk" a disk with no partitions, and it has a GPT label, the system will write the new partition table into the existing label -- it will not wipe out the existing GPT label.

Reply:
But it is not guaranteed that the application will create a partition right. I will list out an example scenario, where I thought partition is required.
1. NDM creates a GPT label on to the block device(`/dev/sdb`) for identification
2. The blockdevice is claimed by an application.
3. The application formats the disk with a filesystem, say ext4. Then the GPT label will be removed right.
This means we no longer can identify the disk.

Reply:
You will have the FSID.

Reply:
But that means, the unique ID we have used has changed(GPT Label -> FSUID). What we need is something that persists irrespective of what the user does.

Reply:
Yes, we need something that is not tied directly to what the user does. But as we know for a very *small* set of cases this is not the truth, unfortunately. 

For cases like these, I think we should follow what the user does, and not assume that the user will never change the FS type, partition layout or anything like that. How will you even prevent the user from not reformatting or repartitioning the disk?

Reply:
Had a chat with @gila and @akhilerm offline on this comment. 

As of today, there is no better option available than to create a partition to identify the virtual disk uniquely across reboots. In case the user after taking the block device completely reformats the disk, there is no way for the NDM to detect and update its resources. This will be added as a known limitation with the current version of NDM.